ignite3/src/main/java/site/ycsb/db/ignite3/AbstractSqlClient.java

```diff
@@ -1,32 +1,100 @@
 package site.ycsb.db.ignite3;
 
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
-import java.util.stream.Collectors;
 import site.ycsb.ByteIterator;
 
 abstract class AbstractSqlClient extends IgniteAbstractClient {
+  /** Prepared statement for reading values. */
+  protected static ThreadLocal<PreparedStatement> readPreparedStatement = new ThreadLocal<>();
 
-  static String prepareInsertStatement(String table, String key, Map<String, ByteIterator> values) {
+  /** Prepared statement for inserting values. */
+  protected static ThreadLocal<PreparedStatement> insertPreparedStatement = new ThreadLocal<>();
+
+  /**
+   * Form a prepared statement SQL command for inserting values.
+   *
+   * @param table Table.
+   * @param values Values.
+   */
+  static String prepareInsertStatement(String table, Map<String, ByteIterator> values) {
     List<String> columns = new ArrayList<>(Collections.singletonList(PRIMARY_COLUMN_NAME));
-    List<String> insertValues = new ArrayList<>(Collections.singletonList(key));
 
     for (Map.Entry<String, ByteIterator> entry : values.entrySet()) {
       columns.add(entry.getKey());
-      insertValues.add(entry.getValue().toString());
     }
 
     String columnsString = String.join(", ", columns);
-    String valuesString = insertValues.stream()
-        .map(e -> "'" + e.replaceAll("'", "''") + "'")
-        .collect(Collectors.joining(", "));
+    String valuesString = String.join(", ", Collections.nCopies(columns.size(), "?"));
 
     return String.format("INSERT INTO %s (%s) VALUES (%s)", table, columnsString, valuesString);
   }
 
+  /**
+   * Init prepared statement object for inserting values.
+   *
+   * @param conn Connection.
+   * @param table Table.
+   * @param values Values.
+   */
+  static PreparedStatement prepareInsertStatement(Connection conn, String table, Map<String, ByteIterator> values)
+      throws SQLException {
+    if (insertPreparedStatement.get() != null) {
+      return insertPreparedStatement.get();
+    }
+
+    insertPreparedStatement.set(conn.prepareStatement(prepareInsertStatement(table, values)));
+
+    return insertPreparedStatement.get();
```

Comment:
```suggestion
    PreparedStatement statement = insertPreparedStatement.get();

    if (statement != null) {
      return statement;
    }

    statement = conn.prepareStatement(prepareInsertStatement(table, values));

    insertPreparedStatement.set(statement);

    return statement;
```

the same for `prepareReadStatement`

Reply:
updated